test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java

```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Test
+    public void givenPrivateClassInPermitsClause_whenCompiling_thenShouldCompile() throws Exception {
+        var root = Path.of("src");
+        tb.writeJavaFiles(root,
+            """
+            sealed class S permits S.A {
+                private static final class A extends S {}
+            }
+            """
+        );
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"))
+            .run(toolbox.Task.Expect.SUCCESS);
+    }
+
+    @Test
+    public void givenPrivateClassOfOtherTopLevelClassInPermitsClause_whenCompiling_thenShouldFail() throws Exception {
+        var root = Path.of("src");
+        tb.writeJavaFiles(root,
+            """
+            public class S {
+                private static final class A extends S {}
+            }
+            """,
+            """
+            public sealed class T permits S.A {
+                private static final class A {}
+            }
+            """
+        );
+        var expectedErrors = List.of(
+            "T.java:1:30: compiler.err.report.access: A, private, S.A",
+            "1 error"
+        );
+
+        var compileErrors = new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"), root.resolve("T.java"))
+            .options("-XDrawDiagnostics")
+            .run(toolbox.Task.Expect.FAIL)
+            .getOutputLines(Task.OutputKind.DIRECT);
+
+        if (!Objects.equals(compileErrors, expectedErrors)) {
+            throw new AssertionError("Expected errors: " + expectedErrors + ", but got: " + compileErrors);
+        }
+
+    }
+
+    @Test
+    public void givenPrivateClassInPermitsClauseOfInnerClass_whenCompiling_thenShouldCompile() throws Exception {
+        var root = Path.of("src");
+        tb.writeJavaFiles(root,
+            """
+            public sealed class S permits S.T.A {
+                static class T {
+                    private static final class A extends S {}
+                }
+            }
+            """
+        );
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"))
+            .run(toolbox.Task.Expect.SUCCESS);
+    }
+
+    @Test
+    public void givenPrivateClassInPermitsClauseContainedInSiblingPrivateInnerClass_whenCompiling_thenShouldCompile() throws Exception {
+        var root = Path.of("src");
+        tb.writeJavaFiles(root,
+            """
+            public class S {
+                private static class A {
+                    private static class B extends C.D {}
+                }
+                private static class C {
+                    private static class D {}
+                }
+            }
+            """
+        );
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"))
+            .run(toolbox.Task.Expect.SUCCESS);
+    }
+
+    @Test
+    public void givenPrivateClassInPermitsClause_whenThanCompilingOtherClassThatReferencesPrivateClassInPermitsClause_thenShouldFail() throws Exception {
```

Comment:
this one is failing for me too

Reply:
It may be. I run heavy machine learning model on y pc right now, so i prefer to run tests using CI to not blow up my PC. I hope this doesnt bother anyone too much so I can figure out correct error messages that are expected in a few iterations.